### PR TITLE
Legend for markers with data-defined size

### PR DESCRIPTION
**Pull request by wonder-sk**
This is a rework of PR #4236 

basic_bottom

Supports:
- single / categorized / graduated symbol renderers for markers. Available only for layers with point geometry. In renderer configuration GUI, click "Advanced" button and select "Data-defined size legend..."
- diagrams with variable diagram size. In diagram configuration GUI, in Legend tab there is a new button "Legend entries for diagram size..."

Data-defined size property in renderers needs to be created using the assistant because it adds a size scale transformer that is necessary to know the expected range of values.

Configuration options:
- separate vs. collapsed legend
- classes - manually set vs. automatically generated using pretty breaks (the only option before)
- for collapsed legend - central vs bottom alignment

image

### CODE REVIEW

**Comment by nirvn** [comment]
Exciting stuff. Does it work with composer legend item too?

**Comment by wonder-sk** [comment]
@nirvn yes :-)

**Comment by haubourg** [comment]
That's wonderful ! Thanks!

**Comment by nirvn** [comment]
@wonder-sk , wonderful.

**Comment by nyalldawson** [comment]
> Data-defined size property in renderers needs to be created using the assistant because it adds a size scale transformer that is necessary to know the expected range of values.

I'm a bit confused by this requirement. Doesn't the new manual size classes option cover this requirement, since users can enter whatever values they want included in the legend?

File: src/core/layertree/qgslayertreemodellegendnode.cpp [hunk 1]
```diff
@@ -215,22 +216,23 @@ void QgsSymbolLegendNode::uncheckAllItems()
...
   if ( model() )
     model()->legendMapViewData( &mupp, &dpi, &scale );
-  bool validData = !qgsDoubleNear( mupp, 0.0 ) && dpi != 0 && !qgsDoubleNear( scale, 0.0 );
+
+  if ( qgsDoubleNear( mupp, 0.0 ) || dpi == 0 || qgsDoubleNear( scale, 0.0 ) )
+    return nullptr;
 
   // setup temporary render context
-  std::unique_ptr<QgsRenderContext> context( new QgsRenderContext );
+  QgsRenderContext *context = new QgsRenderContext;
```
Comment (nyalldawson) [inline src/core/layertree/qgslayertreemodellegendnode.cpp]: I think this should still be a unique_ptr - there should be no cost and it reduces the chance of leaks if someone changes things here in future.

File: tests/src/core/testqgsdatadefinedsizelegend.cpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+  }
+
+  QgsApplication::exitQgis();
+}
+
+void TestQgsDataDefinedSizeLegend::testBasic()
+{
+  QgsDataDefinedSizeLegend settings;
+  settings.setLegendType( QgsDataDefinedSizeLegend::LegendCollapsed );
+  settings.setFont( QgsFontUtils::getStandardTestFont( QStringLiteral( "Roman" ) ) );
```
Comment (nyalldawson) [inline tests/src/core/testqgsdatadefinedsizelegend.cpp]: Can you use a bold font here (and bump up the font size)? it'll be more forgiving to the cross platform rendering differences

File: src/gui/symbology-ng/qgssinglesymbolrendererwidget.cpp [hunk 3]
```diff
@@ -105,3 +114,15 @@ void QgsSingleSymbolRendererWidget::showSymbolLevels()
 {
   showSymbolLevelsDialog( mRenderer );
 }
+
+void QgsSingleSymbolRendererWidget::dataDefinedSizeLegend()
+{
+  bool ok;
+  QgsMarkerSymbol *s = static_cast<QgsMarkerSymbol *>( mSingleSymbol ); // this should be only enabled for marker symbols
+  std::unique_ptr<QgsDataDefinedSizeLegend> ddsLegend( showDataDefinedSizeLegendDialog( s, mRenderer->dataDefinedSizeLegend(), &ok ) );
```
Comment (nyalldawson) [inline src/gui/symbology-ng/qgssinglesymbolrendererwidget.cpp]: No chance of a nice non-blocking QgsPanelWidget here? ;)

**Review by nyalldawson — COMMENTED** [review]
Looks great!

File: src/ui/qgsdatadefinedsizelegendwidget.ui [hunk 4]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <ui version="4.0">
- <class>QgsDataDefinedSizeLegendDialog</class>
- <widget class="QDialog" name="QgsDataDefinedSizeLegendDialog">
+ <class>QgsDataDefinedSizeLegendWidget</class>
+ <widget class="QWidget" name="QgsDataDefinedSizeLegendWidget">
```
Comment (nyalldawson) [inline src/ui/qgsdatadefinedsizelegendwidget.ui]: should this be QgsPanelWidget? not sure if it makes any ultimate difference though.

**Comment by nyalldawson** [comment]
Thanks @wonder-sk !

**Comment by sbrunner** [comment]
Hello,

Thanks for your job, issue I found when I do some tests:

* When I try it many times the legend isn't updated when I change the configuration, e.g. change the diagram size.
* We can create legend like this!

* I find how to set the label of the legend attribute...
* We should be able to change the label of the size, for example if we use the expression `log10(population)` we want to create classes like (value => display) `1 => 1`,` 3 => 1000`, `5 => 100000`, ...

**Comment by sbrunner** [comment]
And for met it would be great the the result of the GetLegendGraphic is tested ...

**Comment by sbrunner** [comment]
When I save and reopen the project the symbol definition is lost ... (I change the symbol color to white, and when I reopen the project it became red...)

**Comment by wonder-sk** [comment]
@sbrunner 

> When I try it many times the legend isn't updated when I change the configuration, e.g. change the diagram size.

Not sure I understand the problem. Is it with diagrams or with renderers using data-defined marker size - or both? What do you need to do for this to happen?

> We can create legend like this!  [picture]

The order of numbers 5 and 7 should be probably different, but otherwise I think the output is correct if you set up size scaling in a way that 5 and 7 have the same size... What output do you expect?

> I find how to set the label of the legend attribute...

Sorry I do not understand - did you mean "I did not find ..." ?

> We should be able to change the label of the size, for example if we use the expression log10(population) we want to create classes like (value => display) 1 => 1,3 => 1000, 5 => 100000, ...

That could be added - maybe better in a follow up PR to get this one merged first.

> And for met it would be great the the result of the GetLegendGraphic is tested ...

The output should be equivalent to what legend renderer test outputs - I would prefer not to add too many test images otherwise it is a pain to update them when necessary

> When I save and reopen the project the symbol definition is lost ... (I change the symbol color to white, and when I reopen the project it became red...)

Should be fixed now.

**Comment by wonder-sk** [comment]
@nyalldawson I have removed the requirement to have size scale transformer associated with data-defined size... it was originally required when manual configuration of classes was not there (so the automatic class creation would not know the range)

**Comment by sbrunner** [comment]
> Not sure I understand the problem. Is it with diagrams or with renderers using data-defined marker size - or both? What do you need to do for this to happen?

* Open the layer properties
* Change the diagram size
* OK

> The order of numbers 5 and 7 should be probably different, but otherwise I think the output is correct if you set up size scaling in a way that 5 and 7 have the same size... What output do you expect?

In the rendered graphic on the map they don't have the same size, than on the legend t should be the same...

> Sorry I do not understand - did you mean "I did not find ..." ?

We should be able to put a text above the diagram size legend to know what that represent, where to configure that?

**Comment by 3nids** [comment]
just tried with graduated (size method) and it says "Data-defined size is not enabled".
Am I missing something?

awesome feature!

**Comment by anitagraser** [comment]
Great stuff!
How about making collapsed legend the default for single symbol renderers when data defined size is enabled? It's quite hidden now behind the advanced button.

**Pull request merged**